packages/clis/parser/game-files/sii-schemas.ts

```diff
@@ -24,42 +27,83 @@ ajv.addKeyword({
 // - can reveal important changes to files that need to be taken into account,
 //   like the changes in icon .mat files to support SDF files.
 
-const NumberTupleSchema: JSONSchemaType<[number, number]> = {
-  type: 'array',
-  items: [{ type: 'number' }, { type: 'number' }],
-  minItems: 2,
-  maxItems: 2,
-};
-const NumberTripleSchema: JSONSchemaType<[number, number, number]> = {
-  type: 'array',
-  items: [{ type: 'number' }, { type: 'number' }, { type: 'number' }],
-  minItems: 3,
-  maxItems: 3,
-};
-const NumberQuadrupleSchema: JSONSchemaType<[number, number, number, number]> =
-  {
+const integer: JSONSchemaType<number> = { type: 'integer' } as const;
+const number: JSONSchemaType<number> = { type: 'number' } as const;
+const string: JSONSchemaType<string> = { type: 'string' } as const;
+// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment
+const bigint: JSONSchemaType<bigint> = { dataType: 'bigint' } as any;
+
+type Tuple<T, N extends number, R extends T[] = []> = R['length'] extends N
+  ? R
+  : Tuple<T, N, [T, ...R]>;
+
+const arrayOf = <T>(items: T) =>
+  ({
     type: 'array',
-    items: [
-      { type: 'number' },
-      { type: 'number' },
-      { type: 'number' },
-      { type: 'number' },
-    ],
-    minItems: 4,
-    maxItems: 4,
-  };
-const StringArraySchema: JSONSchemaType<string[]> = {
-  type: 'array',
-  items: { type: 'string' },
-  minItems: 1,
-};
-const LocaleTokenSchema: JSONSchemaType<string> = {
-  type: 'string',
-  pattern: '^@@.+@@$',
+    items,
+    minItems: 1,
+  }) as const;
+const fixedLengthArray = <T, N extends number>(
+  item: JSONSchemaType<T>,
+  length: N,
+): JSONSchemaType<Tuple<T, N>> =>
+  ({
+    ...arrayOf(item),
+    minItems: length,
+    maxItems: length,
+  }) as unknown as JSONSchemaType<Tuple<T, N>>;
+const stringEnum = <T extends string>(...values: readonly T[]) =>
+  ({
+    type: 'string',
+    enum: values,
+  }) as const;
+const stringPattern = (regex: RegExp) =>
+  ({
+    type: 'string',
+    pattern: regex.source,
+  }) as const;
+const nullable = <T extends object>(t: T) =>
+  ({
+    ...t,
+    nullable: true,
+  }) as const;
+const object = <T extends object, U extends keyof T>(
+  properties: T,
+  // TODO figure out a way to remove the need for this argument. It should be
+  //  calculate-able, based on non-nullable `properties`.
+  required: U[] = Object.keys(properties) as U[],
```

Comment:
This is the third thing I'm not completely happy with. But I've already spent more time than I wanted to wrestling with types.